packages/xo-server-sdn-controller/src/index.js

```diff
@@ -888,23 +887,23 @@ class SDNController extends EventEmitter {
   // ---------------------------------------------------------------------------
 
   async _setPoolControllerIfNeeded(pool) {
-    if (!isControllerNeeded(pool.$xapi)) {
-      // Nothing to do
-      return
-    }
+    if (isControllerNeeded(pool.$xapi)) {
+      const controller = find(pool.$xapi.objects.all, {
+        $type: 'SDN_controller',
+      })
+      if (controller !== undefined) {
+        await pool.$xapi.call('SDN_controller.forget', controller.$ref)
+        log.debug('Old SDN controller removed', {
+          pool: pool.name_label,
+        })
+      }
 
-    const controller = find(pool.$xapi.objects.all, { $type: 'SDN_controller' })
-    if (controller !== undefined) {
-      await pool.$xapi.call('SDN_controller.forget', controller.$ref)
-      log.debug('Old SDN controller removed', {
+      await pool.$xapi.call('SDN_controller.introduce', PROTOCOL)
+      log.debug('SDN controller has been set', {
```

Comment:
The only change here is that you push a cleaner whether the controller was needed or not?

Reply:
Yes I need to call `manageXapi` even if the controller was already set in order to set my callbacks.